test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -1142,58 +1138,103 @@ arbitraryActionWithVars _ ctx _st = QC.frequency $ concat [
+    tableVars =
+      [ fromRight v
+      | v <- findVars ctx Proxy
+      , case lookupVar ctx v of
+          MEither (Left _)                  -> False
+          MEither (Right (MTableHandle th)) ->
+            Map.member (Model.tableHandleID th) (Model.tableHandles st)
+      ]
+
+    genCursorVar = QC.elements cursorVars
+
+    cursorVars :: [Var h (WrapCursor h IO k v blob)]
+    cursorVars =
+      [ fromRight v
+      | v <- findVars ctx Proxy
+      , case lookupVar ctx v of
+          MEither (Left _)            -> False
+          MEither (Right (MCursor c)) ->
+            Map.member (Model.cursorID c) (Model.cursors st)
+      ]
+
+    genBlobRefsVar = QC.elements blobRefsVars
+
+    blobRefsVars :: [Var h (V.Vector (WrapBlobRef h IO blob))]
+    blobRefsVars = fmap (mapGVar (OpComp OpLookupResults)) lookupResultVars
+                ++ fmap (mapGVar (OpComp OpQueryResults))  queryResultVars
       where
-        vars1 = findVars ctx (Proxy @(Either Model.Err (V.Vector (R.LookupResult v (WrapBlobRef h IO blob)))))
-        vars2 = findVars ctx (Proxy @(Either Model.Err (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob)))))
-
-        fromLookupResults ::
-             Var h (Either Model.Err (V.Vector (R.LookupResult v (WrapBlobRef h IO blob))))
-          -> Var h (Either Model.Err (V.Vector (WrapBlobRef h IO blob)))
-        fromLookupResults = mapGVar (\op -> OpRight `OpComp` OpLookupResults `OpComp` OpFromRight `OpComp` op)
-
-        fromQueryResults ::
-             Var h (Either Model.Err (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob))))
-          -> Var h (Either Model.Err (V.Vector (WrapBlobRef h IO blob)))
-        fromQueryResults = mapGVar (\op -> OpRight `OpComp` OpQueryResults `OpComp` OpFromRight `OpComp` op)
-
-    withoutVars :: [(Int, Gen (Any (LockstepAction (ModelState h))))]
-    withoutVars = [
-          (5, Some . New (PrettyProxy :: PrettyProxy (k, v, blob)) <$> QC.arbitrary)
-        , (3, fmap Some $ Open @k @v @blob <$> genSnapshotName)
-        , (1, fmap Some $ DeleteSnapshot <$> genSnapshotName)
-        , (1, pure $ Some ListSnapshots)
+        lookupResultVars :: [Var h (V.Vector (R.LookupResult  v (WrapBlobRef h IO blob)))]
+        queryResultVars  :: [Var h (V.Vector (R.QueryResult k v (WrapBlobRef h IO blob)))]
+
+        lookupResultVars = fromRight <$> findVars ctx Proxy
+        queryResultVars  = fromRight <$> findVars ctx Proxy
+
+    genUsedSnapshotName   = QC.elements usedSnapshotNames
+    genUnusedSnapshotName = QC.elements unusedSnapshotNames
+
+    usedSnapshotNames, unusedSnapshotNames :: [R.SnapshotName]
+    (usedSnapshotNames, unusedSnapshotNames) =
+      partitionEithers
+        [ if Map.member snapshotname (Model.snapshots st)
+            then Left  snapshotname -- used
+            else Right snapshotname -- unused
+        | name <- ["snap1", "snap2", "snap3" ]
+        , let snapshotname = fromJust (R.mkSnapshotName name)
         ]
```

Comment:
It looks to me like this generator by itself could run into an error here, because `QC.elements` on an empty list returns an error. However, I see we only use these generators in combination with checking whether the lists are null, so it should be fine

Reply:
Yeah, it's a bit error-prone in that one has to filter at the use sites based on whether the lists are non-empty. But I don't immediately see a more modular way of doing it. The other ones (tables, cursors) work in this style too.